File: ipbus/src/common/done_counter.cxx
#include <uhal/uhal.hpp>
#include <uhal/definitions.hpp>

#include <vector>
#include <string>
#include <iostream>

#define RDOUT_DONE_MAGIC 0xABCD4321

int main(int argc, char** argv) {
	uhal::ConnectionManager manager("file://etc/connections.xml");
    std::vector<std::string> rdout_ids = manager.getDevices();

    std::vector<uhal::HwInterface> rdouts;
    for(std::vector<std::string>::iterator str = rdout_ids.begin(); str != rdout_ids.end(); str++)
        rdouts.push_back(manager.getDevice(str));

	for(int i = 0; i < 65535; i++) {
        std::vector<uint32_t> count0;
        for(std::vector<uhal::HwInterface>::iterator rdout = rdouts.begin(); rdout != rdouts.end(); rdout++) {
            uhal::ValWord<uint32_t> x = rdout.getNode("RDOUT_DONE_COUNT").read();
            rdout.dispatch();
            count0.push_back(x.value());

            rdout.getNode("RDOUT_DONE").write(RDOUT_DONE_MAGIC);
            rdout.dispatch();
        }

		std::vector<uint32_t> count1;
        for(std::vector<uhal::HwInterface>::iterator rdout = rdouts.begin(); rdout != rdouts.end(); rdout++) {
            uhal::ValWord<uint32_t> x = rdout.getNode("RDOUT_DONE_COUNT").read();
            rdout.dispatch();
            count1.push_back(x.value());
        }

        for(int idx = 0; idx < rdouts.size(); idx++) {
            if(count1[idx]-1 != count0[idx]) {
                std::cout << "bad rdout_done count for " << rdouts[idx].getId() << std::endl;
                return 0;
            }
        }
	}

	return 0;
}

File: ipbus/python/done_counter.py
import uhal
from time import sleep

NUM_SIG = 65535 # max syncboard can count to

uhal.setLogLevelTo(uhal.LogLevel.ERROR)

manager = uhal.ConnectionManager('file://etc/connections.xml')

rdouts = [manager.getDevice(rdout_id) for rdout_id in manager.getDevices()]

for i in range(NUM_SIG):
    count0 = []
    for rdout in rdouts:
        count0.append(rdout.getNode('RDOUT_DONE_COUNT').read())
        rdout.dispatch() # make sure read gets done first
        rdout.getNode('RDOUT_DONE').write(0xABCD4321)
        rdout.dispatch()

    count1 = []
    for rdout in rdouts:
        count1.append(rdout.getNode('RDOUT_DONE_COUNT').read())
        rdout.dispatch()

    for idx in range(len(rdouts)):
        if (count1[idx] - 1) != count0[idx]:
            print 'BAD! RDOUT_DONE_COUNT for', rdouts[idx].id(), 'didn't increment by one!'
            exit()

print 'Done. Counter increments correctly'
